core/src/main/java/org/ehcache/core/StatusTransitioner.java

```diff
@@ -218,7 +218,13 @@ public void succeeded() {
     }
 
     public StateTransitionException failed(Throwable t) {
+      if (st.done()) {
```

Comment:
The point is to repair the status transition only when it failed to.
So if transition was done even due to an `RuntimeException` the `finally` call should not do anything, should act like a noop.
And if the transition is not done, which creates problem of hang, the transition is done and then returned.